File: Java_MVVM/app/src/main/java/eu/gotoinc/requesinjava_mvvm/adapter/UserPagingAdapter.java
package eu.gotoinc.requesinjava_mvvm.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import eu.gotoinc.requesinjava_mvvm.R;
import eu.gotoinc.requesinjava_mvvm.repository.data_models.responses.UserResponse;
import eu.gotoinc.requesinjava_mvvm.util.OnRecyclerItemClickListener;
import eu.gotoinc.requesinjava_mvvm.viewmodel.UserItemViewModel;

public class UserPagingAdapter extends PagedListAdapter<UserResponse, ViewHolder> {
	private OnRecyclerItemClickListener<UserResponse> listener;

	public UserPagingAdapter(@NonNull DiffUtil.ItemCallback<UserResponse> diffCallback) {
		super(diffCallback);
	}

	public UserPagingAdapter(@NonNull DiffUtil.ItemCallback<UserResponse> diffCallback, OnRecyclerItemClickListener<UserResponse> listener) {
		super(diffCallback);
		this.listener = listener;
	}

	@NonNull
	@Override
	public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
		ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.item_list, parent, false);
		return new ViewHolder(binding);
	}

	@Override
	public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
		holder.bind(new UserItemViewModel(getItem(position), listener));
	}

}

File: Kotlin_MVVM/app/src/main/java/eu/gotoinc/kotlinmvvm/adapter/UserPagingAdapter.kt
package eu.gotoinc.kotlinmvvm.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import eu.gotoinc.kotlinmvvm.R
import eu.gotoinc.kotlinmvvm.repository.datamodel.response.UserResponse
import eu.gotoinc.kotlinmvvm.viewmodel.UserItemViewModel

class UserPagingAdapter(
    diffCallback: DiffUtil.ItemCallback<UserResponse>,
    private val onItemClick: (UserResponse?) -> Unit
) : PagedListAdapter<UserResponse, ViewHolder>(diffCallback) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder = ViewHolder(
        DataBindingUtil.inflate(
            LayoutInflater.from(parent.context),
            R.layout.item_list,
            parent,
            false
        )
    )

    override fun onBindViewHolder(holder: ViewHolder, position: Int) =
        holder.bind(UserItemViewModel(getItem(position), onItemClick))

}